cmd/scan/scan.go

```diff
@@ -0,0 +1,289 @@
+	ParentHash  string `json:"parent_hash"`
+	BlockHeight int64  `json:"block_height"`
+}
+
+type Deposit struct {
+	Addr string
+	Tx   Tx
+}
+
+func ScanBlock(client *rpcclient.Client, blockID int64) ([]Deposit, error) {
+	blockHash, err := client.GetBlockHash(blockID)
+	if err != nil {
+		return nil, err
+	}
+	block, err := client.GetBlockVerboseTx(blockHash)
+	if err != nil {
+		return nil, err
+	}
+	parentHash := block.PreviousHash
+	var deposits []Deposit
+	var depTx Tx
+	depTx.BlockHeight = blockID
+	depTx.BlockHash = blockHash.String()
+	depTx.ParentHash = parentHash
+	for _, tx := range block.RawTx {
+		for i, addr := range tx.Vout {
+			depTx.TxHash = fmt.Sprintf("%s:%d", tx.Txid, i)
+
+			if len(addr.ScriptPubKey.Addresses) > 0 {
+				deposits = append(deposits, Deposit{
+					Addr: addr.ScriptPubKey.Addresses[0],
+					Tx:   depTx,
+				})
+			}
+		}
+
+	}
+	return deposits, nil
+}
+
+func CompareAddress(addr Address, deps []Deposit) Address {
+	for _, dep := range deps {
+		if addr.Addr == dep.Addr && !ExistTx(addr, dep.Tx) {
+			addr.Txs = append(addr.Txs, dep.Tx)
+		}
+	}
+
+	return addr
+}
+
+func ExistTx(addr Address, tx Tx) bool {
+	for _, t := range addr.Txs {
+		if t == tx {
+			return true
+		}
+	}
+	return false
+}
+
+func ExistAddress(newAddr Address, walletAddresses []Address) bool {
+	for _, addr := range walletAddresses {
+		if newAddr.Addr == addr.Addr {
+			return true
+		}
+	}
+	return false
+}
+
+func UpdateAddressInfo(addrs []Address, deps []Deposit, blockID int64) []Address {
+
+	for i, addr := range addrs {
+		switch {
+		case addr.MaxScanBlock == 0 && blockID > 1:
+			addr = CompareAddress(addr, deps)
+			addrs[i].Txs = addr.Txs
+			addrs[i].MaxScanBlock = blockID
+			addrs[i].MidScanBlock = blockID
+
+		case addr.MinScanBlock < addr.MaxScanBlock && addr.MinScanBlock == (blockID-1):
+			addr = CompareAddress(addr, deps)
+			addrs[i].Txs = addr.Txs
+			addrs[i].MinScanBlock = blockID
+			if addrs[i].MinScanBlock > addrs[i].MidScanBlock {
+				addrs[i].MidScanBlock = addrs[i].MinScanBlock
+			}
+		case addr.MaxScanBlock > addr.MinScanBlock && addr.MaxScanBlock == (blockID-1):
+			addr = CompareAddress(addr, deps)
+			addrs[i].Txs = addr.Txs
+			addrs[i].MaxScanBlock = blockID
+		case addr.MinScanBlock == addr.MidScanBlock && addr.MinScanBlock == addr.MaxScanBlock && addr.MaxScanBlock == (blockID-1):
+			addr = CompareAddress(addr, deps)
+			addrs[i].Txs = addr.Txs
+			addrs[i].MaxScanBlock = blockID
+			addrs[i].MidScanBlock = blockID
+			addrs[i].MinScanBlock = blockID
+		}
+
+	}
+	return addrs
+}
+
+func LoadWallet(file string) ([]Address, error) {
+	var addrs []Address
+	wallet, err := os.Open(file)
+	defer wallet.Close()
+	if err != nil {
+		return nil, err
+	}
+	jsonParser := json.NewDecoder(wallet)
+	err = jsonParser.Decode(&addrs)
+	if err != nil {
+		return nil, err
+	}
+	return addrs, nil
+}
+
+func SaveWallet(file string, addrs []Address) error {
+	wallet, err := os.Open(file)
+	defer wallet.Close()
```

Comment:
defer close after check error

Reply:
Need to check Close() error for writable files, see
https://joeshaw.org/dont-defer-close-on-writable-files/

Reply:
thanks for article